### DISCUSSION TITLE
Add server hooks

### DISCUSSION THREAD

**Discussion by abersheeran** [Kludex/uvicorn#1936] [Ideas]
Similar to [Gunicorn](https://docs.gunicorn.org/en/stable/settings.html#server-hooks).

This will solve more problems for us. For example https://stackoverflow.com/questions/58133694/graceful-shutdown-of-uvicorn-starlette-app-with-websockets . If we provide a `before_shutdown` hook, we can handle it gracefully without dangerously modifying uvicorn's internal code.

This is just an idea at the moment.

@encode/maintainers Do anybody have any thoughts?

**Comment by Kludex** [comment]
If we add server hooks, like `post_request`, we can use it for custom access log messages.

**Reply by abersheeran** [reply]
Yeah.

**Comment by Kludex** [comment]
@tomchristie What do you think?

**Reply by lovelydinosaur** [reply]
I'm unclear what functionality this supports that isn't already covered by the ASGI messaging.

(I might not be the best person to ask tho)
